modules/renter.go

```diff
@@ -50,12 +54,92 @@ type Allowance struct {
-	Received    uint64    `json:"received"`
-	StartTime   time.Time `json:"starttime"`
-	Error       string    `json:"error"`
+	SiaPath     string         `json:"siapath"`
+	Destination DownloadWriter `json:"destination"`
+	Filesize    uint64         `json:"filesize"`
+	Received    uint64         `json:"received"`
+	StartTime   time.Time      `json:"starttime"`
+	Error       string         `json:"error"`
+}
+
+// DownloadWriter provides an interface which all output writers have to implement.
+type DownloadWriter interface {
+	WriteAt(b []byte, off int64) (int, error)
+}
+
+// DownloadFileWriter is a file-backed implementation of DownloadWriter.
+type DownloadFileWriter struct {
+	f        *os.File
+	Location string
+	offset   uint64
+}
+
+// NewDownloadFileWriter creates a new instance of a DownloadWriter backed by the file named.
+func NewDownloadFileWriter(fname string, offset, length uint64) *DownloadFileWriter {
+	l, _ := os.OpenFile(fname, os.O_CREATE|os.O_WRONLY, defaultFilePerm)
+	return &DownloadFileWriter{
+		f:        l,
+		Location: fname,
+		offset:   offset,
+	}
+}
+
+// WriteAt writes the passed bytes at the specified offset.
+func (dw *DownloadFileWriter) WriteAt(b []byte, off int64) (int, error) {
+	fileOffset := off - int64(dw.offset)
+
+	r, err := dw.f.WriteAt(b, fileOffset)
+	if err != nil {
+		build.ExtendErr("unable to write to download destination", err)
+	}
+	dw.f.Sync()
+
+	return r, err
+}
+
+// DownloadHttpWriter is a http response writer-backed implementation of DownloadWriter.
+// The writer writes all content that is written to the current `offset` directly to the ResponseWriter,
+// and buffers all content that is written at other offsets.
+// After every write to the ResponseWriter the `offset` and `length` fields are updated, and buffer content written until
+type DownloadHttpWriter struct {
+	w              http.ResponseWriter
+	offset         int         // The index in the original file of the last byte written to the response writer.
+	firstByteIndex int         // The index of the first byte in the original file.
+	length         int         // The total size of the slice to be written.
+	buffer         []byte      // Buffer used for storing the chunks until download finished.
+	offsets        map[int]int // Map from offset to length of chunk.
+}
+
+// NewDownloadHttpWriter creates a new instance of http.ResponseWriter backed DownloadWriter.
+func NewDownloadHttpWriter(w http.ResponseWriter, offset, length uint64) *DownloadHttpWriter {
+	return &DownloadHttpWriter{
+		w:              w,
+		offset:         int(offset),
+		firstByteIndex: int(offset),
+		length:         int(length),
+		buffer:         make([]byte, length),
+		offsets:        make(map[int]int),
+	}
+}
+
+// WriteAt buffers parts of the file until the entire file can be
+// flushed to the client. Returns the number of bytes written or an error.
+func (dw *DownloadHttpWriter) WriteAt(b []byte, off int64) (int, error) {
```

Comment:
In the worst case, this will result in buffering the entire file in memory. This might work fine for downloading a single chunk (although chunks can still be quite large), but later someone else might come along and assume that `DownloadHttpWriter` could be used to stream a whole file.

More generally: there is a lot of code sharing going on here that makes me uneasy. I think `Download` and `DownloadSection` are fundamentally different operations, so sharing code between them is a bonus, not a requirement.
Perhaps we could have some function that downloaded a chunk and held it in memory, and then `Download` would write it out to disk and `DownloadSection` would write it out to the HTTP response. But if the current renter design makes that difficult or impractical, I'd rather we just write the `DownloadSection` logic separately. Long-term, we will want to optimize `DownloadSection` in ways that aren't applicable to `Download`. And I think @DavidVorick is itching to redesign the download code anyway, so all the more reason to keep them separate.